app/Traits/TwoStep.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * NOTICE OF LICENSE
+ *
+ * UNIT3D is open-sourced software licensed under the GNU General Public License v3.0
+ * The details is bundled with this project in the file LICENSE.txt.
+ *
+ * @project    UNIT3D
+ * @license    https://choosealicense.com/licenses/gpl-3.0/  GNU General Public License v3.0
+ * @author     HDVinnie
+ */
+
+namespace App\Traits;
+
+use Auth;
+use Carbon\Carbon;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Validator;
+use App\TwoStepAuth;
+use App\Notifications\TwoStepAuthCode;
+
+trait TwoStep
+{
+    /**
+     * Check if the user is authorized
+     *
+     * @param  Request $request
+     *
+     * @return boolean
+     */
+    public function twoStepVerification($request)
+    {
+        $user = Auth::User();
+
+        if ($user) {
+
+            $twoStepAuthStatus = $this->checkTwoStepAuthStatus($user->id);
+
+            if ($twoStepAuthStatus->authStatus !== true) {
+                return false;
+            }
+            if ($this->checkTimeSinceVerified($twoStepAuthStatus)) {
+                return false;
+            }
+
+            return true;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check time since user was last verified and take apprpriate action
+     *
+     * @param collection $twoStepAuth
+     *
+     * @return boolean
+     */
+    private function checkTimeSinceVerified($twoStepAuth)
+    {
+        $expireMinutes  = config('auth.TwoStepVerifiedLifetimeMinutes');
+        $now            = Carbon::now();
+        $expire         = Carbon::parse($twoStepAuth->authDate)->addMinutes($expireMinutes);
+        $expired        = $now->gt($expire);
+
+        if ($expired) {
+            $this->resetAuthStatus($twoStepAuth);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Reset TwoStepAuth collection item and code
+     *
+     * @param collection $twoStepAuth
+     *
+     * @return collection
+     */
+    private function resetAuthStatus($twoStepAuth)
+    {
+        $twoStepAuth->authCode    = $this->generateCode();
+        $twoStepAuth->authCount   = 0;
+        $twoStepAuth->authStatus  = 0;
+        $twoStepAuth->authDate    = NULL;
+        $twoStepAuth->requestDate = NULL;
+
+        $twoStepAuth->save();
+
+        return $twoStepAuth;
+    }
+
+    /**
+     * Generate Authorization Code
+     *
+     * @param integer $length
+     * @param string $prefix
+     * @param string $suffix
+     *
+     * @return string
+     */
+    private function generateCode(int $length = 4, string $prefix = '', string $suffix = '')
+    {
+        for($i = 0; $i < $length; $i++){
+            $prefix .= random_int(0,1) ? chr(random_int(65, 90)) : random_int(0, 9);
```

Comment:
This might be better if you have an array with codeChars, which enumerates the chars allowed, some people might want to change this later in the config. Then you could replace it then with:
```php
$prefix = codeChars[random_int(0, count(codeChars-1)];
```